### DISCUSSION TITLE
glass-server reports no available dbs

### DISCUSSION THREAD

**Discussion by xvandish** [facebookincubator/Glean#451] [Q&A]
Hello!

Caveat: I've barely poked through the code, I will do, but putting this up in case it's an obvious answer. I'm trying to figure out if/what I'm doing is wrong, any help appreciated, thank you!

I built Glean from source on an Ubuntu x86 image, indexed some demo code, queried the db that was created a bit, and then was trying to experiment with the `glass-server`. However, when starting up the server, it reported no available db's, and using the `glass-democlient`, the server reports `repo not found` for the repo whose' db exists. 

I tried pointing `glass-server` to a running `glean server`, and also by pointing to a `db-root & schema`

`glean server` did see the db, and I could query/list it using `glean shell`.

## glean server finds the db fine
```
....:~$ exec glean-server --db-root glean-dbs --schema Glean/glean/schema/source/ --port 12345 &                                                                                                           
[1] 139864
....:~$ I0602 09:16:33.072455 139864 Server.hs:94] Starting server
I0602 09:16:33.072866 139864 Server.hs:102] server alive on port 12345
I0602 09:19:05.695835 139868 Open.hs:493] zoekt/0: opening
I0602 09:19:05.719218 139868 Open.hs:499] zoekt/0: opened
I0602 09:19:06.044571 139868 Open.hs:501] zoekt/0: schema has 1084 predicates
```

## glean shell finds the db fine
```
...:~$ glean shell --db-root glean-dbs/                                                                                                                                                                 [151/9783]
Glean Shell, built on 2024-06-02 07:36:16.339322769 UTC, from rev 2b6866e419481cffebae90e77535a970ebd1ac9c
Using local DBs from rocksdb:glean-dbs/
type :help for help.
> :list
zoekt/0 (complete)
  Created: 2024-06-02 08:14:20 UTC (17 minutes ago)
  Completed: 2024-06-02 08:14:37 UTC (17 minutes ago)

>
```

## created an index out of the schema for glass-server
```
# created an index and put it where glass-server wants it
...:~/Glean$ ./dist-newstyle/build/x86_64-linux/ghc-8.10.7/glean-0.1.0.0/x/gen-schema/build/gen-schema/gen-schema --dir glean/schema/source --update-index ~/.config/glean/schema-index
parse/resolve/typecheck: 608.49ms, 1.10 GiB
checking schema roundtrip: 196.53ms, 355.37 MiB
Updating index in /home/ubuntu/.config/glean/schema-index
New index contains 1 instance(s)
gen: 2.01us, 336 bytes
```

## glass-server finds no db's, using db-root
```
# glass-server pointing to db-root & schema
glass-server --db-root glean-dbs/ --schema dir:Glean/glean/schema/source
I0602 09:00:59.494324 139524 Repos.hs:290] filtered available DBs in 610.00ns, 0/0 unavailable
I0602 09:00:59.494554 139524 Repos.hs:380] fetched commit generations for 0 DBs in 127.05us
I0602 09:00:59.494648 139524 Main.hs:131] glass: port 26073, config glean/glass/glass_server
```

## glass-server finds no db's, pointing to a glean-server
```
# example pointing to a glean server running on :12345
glass-server --service localhost:12345
I0602 09:16:37.392989 139873 Repos.hs:290] filtered available DBs in 620.00ns, 0/0 unavailable
I0602 09:16:37.393266 139873 Repos.hs:380] fetched commit generations for 0 DBs in 181.21us
I0602 09:16:37.393373 139873 Main.hs:131] glass: port 26073, config glean/glass/glass_server
```

## In both cases the `glass-democlient` reports back, `repo not found`
```
...:~$ ./dist-newstyle/build/x86_64-linux/ghc-8.10.7/glean-0.1.0.0/x/glass-democlient/build/glass-democlient/glass-democlient list zoekt/bits.go
glass-democlient: ServerException {serverException_message = "No repository found for: zoekt (go)"}
```

**Comment by xvandish** [comment]
Also, thanks for open sourcing Glean and for all the work you put into it, it's very cool stuff!

**Comment by simonmar** [comment]
Sorry for the delay here, I think this question slipped through the net. Glass has a static list of DBs that it knows about, which is configured in `Glean.Glass.RepoMapping`: https://github.com/facebookincubator/Glean/blob/main/glean/glass/Glean/Glass/RepoMapping.hs#L43

To get your DBs to be recognised, edit the list and recompile.

Really this should be done in a better way with some runtime configuration.
